[package] (Un)package GraphModules using corresponding Tracer

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57221 [package] (Un)package GraphModules using corresponding Tracer**
* #57220 [fx] Make Tracer._autowrap_search a list of module names

**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

Differential Revision: [D28164471](https://our.internmc.facebook.com/intern/diff/D28164471)

Commit:
[package] (Un)package GraphModules using corresponding Tracer

**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

[ghstack-poisoned]

SplitInfinity commented on test/package/test_package_fx.py:
```diff
@@ -17,19 +17,31 @@
     from common import PackageTestCase
 
 
+class CustomTracer(Tracer):
+    def is_leaf_module(self, m, module_qualified_name,) -> bool:
+        return super().is_leaf_module(m, module_qualified_name) or module_qualified_name == "MyModule"
+
+
+class MyModule(torch.nn.Module):
+    def forward(self, x):
+        return torch.relu(x)
+
+
+class SimpleTest(torch.nn.Module):
+    def forward(self, x):
+        return torch.relu(x + 3.0)
```
I don't know why, but these are not pickleable when defined in local scope (also why are `MyModule` and `SimpleTest` being pickled in the first place?).

SplitInfinity commented on test/package/test_package_fx.py:
```diff
@@ -17,19 +17,31 @@
 
         f = BytesIO()
         with PackageExporter(f, verbose=False) as pe:
+            pe.extern(["io", "sys"])
```
I think these are needed in order to pickle the tracer?

suo replied:
This comment (and the one about MyModule and SimpleTest being pickled) are vaguely concerning. Can you look at the package to see what's happening? I suspect we are packaging this test file which is not what we want; we should move the custom tracer and other test modules into their own modules if that's the case

SplitInfinity commented on test/package/test_package_fx.py:
```diff
@@ -116,6 +128,33 @@ def test_package_fx_with_imports(self):
+
+        # TODO: How to check that loaded uses CustomTracer (rather, packaged
+        # CustomTracer)?
+        # self.assertIsInstance(loaded.graph.tracer, CustomTracer)
```
So this check obviously doesn't work because the loaded `CustomTracer` is a distinct class with the same name. What can I do instead?

suo replied:
stash an attribute on the custom tracer that doesn't exist on normal tracers and make sure that it matches?

SplitInfinity commented on torch/fx/symbolic_trace.py:
```diff
@@ -431,7 +431,11 @@ def forward(*args, **kwargs):
         self.submodule_paths = None
 
-        return self.graph
+        graph = self.graph
+        graph.tracer = self
+        self.graph = None
```
I ran into pickling issues due to this cyclical reference so I added this to break the cycle when tracing is done.

suo replied:
seems fine but add a comment?

suo replied:
actually what was the error? circular references should be supported by the pickler.

SplitInfinity replied:
Something something `SimpleTest` could not be pickled. I don't have conclusive proof that it was because of this circular reference, but I did notice that pickling works fine without it.

SplitInfinity replied:
After making other changes (like moving those test classes into a separate module) I'm getting a different error. I uploaded a new version of the PR with all the other changes but I'll keep looking into this.

SplitInfinity replied:
Wait actually, we intentionally do not package the graph in `__reduce_package__` for `GraphModule`:
```
    def __reduce_package__(self, exporter: PackageExporter):
        generated_module_name = f'fx-generated._{exporter.get_unique_id()}'
        python_code = self.recompile()
        import_block = _format_import_block(python_code.globals, exporter.importer)
        module_code = import_block + self.code
        exporter.save_source_string(generated_module_name, module_code)

        if hasattr(self._graph, 'tracer'):
            self.tracer = self._graph.tracer

        dict_without_graph = self.__dict__.copy()
        del dict_without_graph['_graph']
        return (reduce_package_graph_module, (dict_without_graph, generated_module_name))
```
If we *don't* break this reference cycle, we end up pickling the graph along with the tracer, which might be causing this problem.

suo replied:
okay that makes sense. I wonder if we should add some basic enforcement to `Graph` so that it just throws if you try to pickle it. I can imagine that custom tracer authors may keep the graph around for whatever reason.

SplitInfinity commented on test/package/test_package_fx.py:
```diff
@@ -17,19 +17,31 @@
 
 
+class CustomTracer(Tracer):
+    def is_leaf_module(self, m, module_qualified_name,) -> bool:
+        return super().is_leaf_module(m, module_qualified_name) or module_qualified_name == "MyModule"
```
Is there anything else that's simple yet unique that I can do in this tracer to make it super obvious when a `Graph` is produced using it? I think this does basically the same thing as the `KeepModules` tracer from `GraphModule` deserialization, so I don't think the graphs produced by the two would differ.

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

[ghstack-poisoned]

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

[ghstack-poisoned]

suo commented on torch/fx/graph_module.py:
```diff
@@ -95,17 +95,23 @@ def __init__(self, body):
+    com = CodeOnlyModule(body)
 
-    from .symbolic_trace import Tracer
+    if 'tracer' not in body:
```
maybe add a note that this is strictly for BC at this point.

SplitInfinity replied:
Graphs can be constructed by hand too, right? Such graphs won't have a tracer.

But yes this is mostly for BC.

suo commented on torch/fx/graph_module.py:
```diff
@@ -520,6 +526,9 @@ def __reduce_package__(self, exporter: PackageExporter):
         module_code = import_block + self.code
         exporter.save_source_string(generated_module_name, module_code)
 
+        if hasattr(self._graph, 'tracer'):
```
I don't think this this should be a side effect of the `__reduce_package__` code. Let's just add it to the init of of `GraphModule` instead, so that all GraphModules retain a reference to their originating tracer.

SplitInfinity replied:
By raising an `Exception` in `__getstate__/__setstate__`?

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

[ghstack-poisoned]

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

Differential Revision: [D28164471](https://our.internmc.facebook.com/intern/diff/D28164471)

[ghstack-poisoned]

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

Differential Revision: [D28164471](https://our.internmc.facebook.com/intern/diff/D28164471)

[ghstack-poisoned]

SplitInfinity commented:
I addressed all comments but I expect a mypy failure due to `Graph` not having a `tracer` attribute. `Tracer` lives in `symbolic_trace.py` which depends on `graph.py`, so there is a circular dependency here. Will a quoted type annotation work:
```
self.tracer: Optional["Tracer"] = None
```
I've only done this in the definition of the class being used in the quoted annotation.

suo commented on torch/fx/symbolic_trace.py:
```diff
@@ -432,7 +432,14 @@ def forward(*args, **kwargs):
+        # Since the tracer might be packaged along with a GraphModule that it helped generate, and that
+        # graphs are explicilty not saved, set self.graph to None so that it doesn't get packaged
+        # along with the tracer.
+        graph = self.graph
```
This isn't really actionable for this PR, but we should clarify what the expected tracer semantics are (and give guidance to custom tracer writers). In particular, we should make it clear that `trace()` should not permanently mutate tracer state, so that we can guarantee that a tracer can always be used to reconstruct a graphmodule even across a serialization boundary.

Looks like right now we informally enforce this (we blank out some of the scratch state like `submodule_paths` and reinitialize some other state like `self.root` and `self.graph`) but it would be good to have a documented expectation. @jamesr66a what do you think?

suo commented:
yeah the quoted type annotation should work

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

Differential Revision: [D28164471](https://our.internmc.facebook.com/intern/diff/D28164471)

[ghstack-poisoned]

bradleyhd commented:
General comment: can we get a clear contract for which classes will have a copy of the graph and under which conditions? It's not entirely clear especially with these changes when a Tracer will have a copy of the graph, and it appears folks have already discovered tracer.graph and rely on it 😅 (Same may end up happening when GraphModules only sometimes have tracers attached to them)

bradleyhd commented on torch/fx/graph.py:
```diff
@@ -279,6 +279,7 @@ def __init__(self, owning_module: Optional["GraphModule"] = None):
         self._graph_namespace = _Namespace()
         self._owners = 0
         self._owning_module = owning_module
+        self.graph: Optional['Tracer'] = None
```
what is this attribute for? it's called `graph` but the type is a `Tracer`?

suo commented:
@bradleyhd I agree, and it relates to my comment above: I don't think we should be stashing any state related to a single trace on the `Tracer`, since we may use a tracer multiple times on different modules or during deserialization.

Right now, we have a "soft" version of that contract, since `trace()` tends to blank out any trace-specific state before starting. But as you pointed out, we are leaving some stuff on the `Tracer` that we probably shouldn't.

One option is to put all the trace-local state into a `TracingContext` object/dict that lives on `Tracer`. That way we can still conveniently access tracing state by doing `self.tracing_context.graph` or whatever, but there's only one thing we need to blank out at the end of tracing.

Commit:
Update on "[package] (Un)package GraphModules using corresponding Tracer"


**Summary**
This commit updates how a `GraphModule` is packaged to pickle and
package the tracer used to produce said `GraphModule` (technically the
`Graph` it owns) and to unpackage it and use it to reconstruct the
`GraphModule` (as opposed to the vanilla tracer).

**Test Plan**
This commit adds `test_package_fx_custom_tracer`, a unit test that
traces a module with a custom tracer and packages it.

**Fixes**
This commit fixes #54573.

Differential Revision: [D28164471](https://our.internmc.facebook.com/intern/diff/D28164471)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28164471).

jamesr66a commented:
> One option is to put all the trace-local state into a TracingContext object/dict that lives on Tracer. That way we can still conveniently access tracing state by doing self.tracing_context.graph or whatever, but there's only one thing we need to blank out at the end of tracing.

SGTM
